File: branches/feature/Android_profile_service/src/org/openremote/android/console/view/ScreenView.java
/*
 * OpenRemote, the Home of the Digital Home. Copyright 2008-2009, OpenRemote Inc.
 * 
 * See the contributors.txt file in the distribution for a full listing of individual contributors.
 * 
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General
 * Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU Affero General Public License along with this program. If not, see
 * <http://www.gnu.org/licenses/>.
 */
package org.openremote.android.console.view;

import java.io.FileNotFoundException;
import java.util.ArrayList;

import org.openremote.android.console.bindings.Background;
import org.openremote.android.console.bindings.LayoutContainer;
import org.openremote.android.console.bindings.Screen;
import org.openremote.android.console.model.PollingHelper;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.util.Log;
import android.widget.AbsoluteLayout;
import android.widget.ImageView;

public class ScreenView extends AbsoluteLayout {

   private Screen screen;
   private PollingHelper polling;
   private AsyncLoader pollingTask;
   /**
    * Instantiates a new screen view.
    * 
    * @param context
    *           the context
    * @param screen
    *           the screen
    */
   @SuppressWarnings("deprecation")
   public ScreenView(Context context, Screen screen) {
      super(context);
      this.screen = screen;
      setBackgroundColor(0);
      setTag(screen.getName());
      setId(screen.getScreenId());
      if (screen.getBackground() != null) {
         addBackground();

      }

      ArrayList<LayoutContainer> layouts = screen.getLayouts();
      for (int i = 0; i < layouts.size(); i++) {
         LayoutContainerView la = LayoutContainerView.buildWithLayoutContainer(context, layouts.get(i));
         if (la != null) {
            LayoutContainer layout = layouts.get(i);
            addView(la, new AbsoluteLayout.LayoutParams(layout.getWidth(), layout.getHeight(), layout.getLeft(), layout
                  .getTop()));
         }
      }
      if (!screen.getPollingComponentsIds().isEmpty()) {
         polling = new PollingHelper(screen.getPollingComponentsIds(), getContext());
      }
   }

   /**
    * @param screen
    */
   private void addBackground() {
      ImageView backgroudView = new ImageView(this.getContext());
      int left = 0;
      int top = 0;
      int screenWidth = Screen.SCREEN_WIDTH;
      int screenHeight = Screen.SCREEN_HEIGHT - Screen.SCREEN_STATUS_BAR_HEIGHT;
      try {
         Bitmap backgroudBitMap = BitmapFactory.decodeStream(this.getContext().openFileInput(screen.getBackgroundSrc()));
         backgroudView.setImageBitmap(backgroudBitMap);
         if (backgroudBitMap == null) {
            return;
         }
         int imageWidth = backgroudBitMap.getWidth();
         int imageHeight = backgroudBitMap.getHeight();
         Background background = screen.getBackground();
         if (!background.isFillScreen()) {
            if (background.isBackgroundImageAbsolutePosition()) {
               left = background.getBackgroundImageAbsolutePositionLeft();
               top = background.getBackgroundImageAbsolutePositionTop();
            } else {
               String backgroundImageRelativePosition = background.getBackgroundImageRelativePosition();
               if ("top-left".equals(backgroundImageRelativePosition)) {
               } else if ("top".equals(backgroundImageRelativePosition)) {
                  left = (screenWidth - imageWidth) / 2;
               } else if ("top-right".equals(backgroundImageRelativePosition)) {
                  left = screenWidth - imageWidth;
               } else if ("left".equals(backgroundImageRelativePosition)) {
                  top = (screenHeight - imageHeight) / 2;
               } else if ("center".equals(backgroundImageRelativePosition)) {
                  left = (screenWidth - imageWidth) / 2;
                  top = (screenHeight - imageHeight) / 2;
               } else if ("right".equals(backgroundImageRelativePosition)) {
                  left = screenWidth - imageWidth;
                  top = (screenHeight - imageHeight) / 2;
               } else if ("bottom".equals(backgroundImageRelativePosition)) {
                  left = (screenWidth - imageWidth) / 2;
                  top = screenHeight - imageHeight;
               } else if ("bottom-left".equals(backgroundImageRelativePosition)) {
                  top = screenHeight - imageHeight;
               } else if ("bottom-right".equals(backgroundImageRelativePosition)) {
                  left = screenWidth - imageWidth;
                  top = screenHeight - imageHeight;
               }
            }
         }
         addView(backgroudView, new AbsoluteLayout.LayoutParams(imageWidth, imageHeight, left, top));
      } catch (FileNotFoundException e) {
         Log.e("ScreenView", "screen background file" + screen.getBackgroundSrc() + " not found.", e);
      }
   }

   public void startPolling() {
      if (polling != null) {
         pollingTask = new AsyncLoader();
         pollingTask.execute((Void)null);
      }
   }
   
   public void cancelPolling() {
      if (polling != null) {
         polling.cancelPolling();
         if (pollingTask != null) {
            pollingTask.cancel(true);
         }
      }
   }
   
   public Screen getScreen() {
      return screen;
   }
   
   class AsyncLoader extends AsyncTask<Void, Void, Integer>{
      @Override
      protected Integer doInBackground(Void... params) {
         polling.requestCurrentStatusAndStartPolling();
         return null;
      }
   }
}

File: attic/trunk/Panels/iOS/Classes/ScreenView.m
/*
 * OpenRemote, the Home of the Digital Home.
 * Copyright 2008-2011, OpenRemote Inc.
 *
 * See the contributors.txt file in the distribution for a
 * full listing of individual contributors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#import "ScreenView.h"
#import "Control.h"
#import "ControlView.h"
#import "LayoutContainer.h"
#import "LayoutContainerView.h"
#import "ServerDefinition.h"
#import "ViewHelper.h"
#import "DirectoryDefinition.h"
#import "XMLEntity.h"
#import "ClippedUIImage.h"
#import "UIViewUtil.h"

@interface ScreenView (Private) 
- (void)createLayout;
- (void)layoutBackground;
@end

@implementation ScreenView

@synthesize screen;

//override the constractor
- (id)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
			[self setBackgroundColor:[UIColor blackColor]];
    }
    return self;
}

//Set screen *and* render its layout 
- (void)setScreen:(Screen *)s{
	[s retain];
	[screen release];
	screen = s;
	[self createLayout];
}
	
//create each layout container in screen
- (void)createLayout {
	[self layoutBackground];
	
	for (LayoutContainer *layout in screen.layouts) {
		[self addSubview:[LayoutContainerView layoutContainerViewWithLayoutContainer:layout]];
	}

}

//override layoutSubviews method of UIView, In order to resize the ControlView when add subview
// Only in this time we can know this view's size
- (void)layoutSubviews {
}

#pragma mark Private method

// Render backgound of ScreenView.
- (void)layoutBackground {
	if ([[[screen background] backgroundImage] src] && [[NSFileManager defaultManager] fileExistsAtPath:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:[[[screen background] backgroundImage] src]]]) {
		UIImage *backgroundImage = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:[[[screen background] backgroundImage] src]]];
		int screenBackgroundImageViewWidth = 0;
		int screenBackgroundImageViewHeight = 0;
		
		if (screen.landscape) {
			screenBackgroundImageViewWidth = [UIScreen mainScreen].bounds.size.height;
			screenBackgroundImageViewHeight = [UIScreen mainScreen].bounds.size.width;
		} else {
			screenBackgroundImageViewWidth = [UIScreen mainScreen].bounds.size.width;
			screenBackgroundImageViewHeight = [UIScreen mainScreen].bounds.size.height;
		}
		//screenBackgroundImageViewHeight -= IPHONE_SCREEN_BOTTOM_PAGE_CONTROL_HEIGHT;

		UIImageView *backgroundImageView = [[[UIImageView alloc] init] autorelease];
		// fillscreen is false
		if (![[screen background] fillScreen]) {
			NSLog(@"BackgroundImage isn't fillScreen");
			NSLog(@"BackgroundImage's original width:%f, height:%f", backgroundImage.size.width, backgroundImage.size.height);
			
			// ablosute position of screen background.
			if ([[screen background] isBackgroundImageAbsolutePosition]) {
				int left = [[screen background] backgroundImageAbsolutePositionLeft];
				int top = [[screen background] backgroundImageAbsolutePositionTop];
				if (left > 0) {
					screenBackgroundImageViewWidth = screenBackgroundImageViewWidth-left;
				}
				if (top > 0) {
					screenBackgroundImageViewHeight = screenBackgroundImageViewHeight-top;
				}
				[backgroundImageView setFrame:CGRectMake(left, top, screenBackgroundImageViewWidth, screenBackgroundImageViewHeight)];
				backgroundImageView = [UIViewUtil clippedUIImageViewWith:backgroundImage dependingOnUIView:backgroundImageView uiImageAlignToUIViewPattern:IMAGE_ABSOLUTE_ALIGN_TO_VIEW isUIImageFillUIView:YES];
				if (left < 0) {
					left = 0;
				}
				if (top < 0) {
					top = 0;
				}
				[backgroundImageView setFrame:CGRectMake(left, top, backgroundImageView.frame.size.width, backgroundImageView.frame.size.height)];
				NSLog(@"Clipped BackgroundImage's width:%f, height:%f", backgroundImageView.image.size.width, backgroundImageView.image.size.height);
				NSLog(@"BackgroundImageView's left is %d, top is %d", left, top);
				NSLog(@"BackgroundImageView's width:%f, height:%f", backgroundImageView.frame.size.width, backgroundImageView.frame.size.height);
			}
			// relative position of screen background.
			else {
				// relative position
				[backgroundImageView setFrame:CGRectMake(0, 0, screenBackgroundImageViewWidth, screenBackgroundImageViewHeight)];
				NSString *backgroundImageRelativePosition = [[screen background] backgroundImageRelativePosition];
				backgroundImageView = [UIViewUtil clippedUIImageViewWith:backgroundImage dependingOnUIView:backgroundImageView uiImageAlignToUIViewPattern:backgroundImageRelativePosition isUIImageFillUIView:NO];
			}
		}
		// fillscreen is true
		else {
			[backgroundImageView setFrame:CGRectMake(0, 0, screenBackgroundImageViewWidth, screenBackgroundImageViewHeight)];
			backgroundImageView = [UIViewUtil clippedUIImageViewWith:backgroundImage dependingOnUIView:backgroundImageView uiImageAlignToUIViewPattern:IMAGE_ABSOLUTE_ALIGN_TO_VIEW isUIImageFillUIView:YES];
		}
		NSLog(@"Added width: %d, height: %d backgroundImageView", screenBackgroundImageViewWidth, screenBackgroundImageViewHeight);
		[self addSubview:backgroundImageView];
		[backgroundImage release];
	}
}

- (void)dealloc {
	[screen release];
	[super dealloc];
}


@end
